nodes.py

```diff
@@ -32,27 +32,55 @@ def gguf_sd_loader_get_orig_shape(reader, tensor_name):
         raise TypeError(f"Bad original shape metadata for {field_key}: Expected ARRAY of INT32, got {field.types}")
     return torch.Size(tuple(int(field.parts[part_idx][0]) for part_idx in field.data))
 
-def gguf_sd_loader(path):
+def gguf_sd_loader(path, handle_prefix="model.diffusion_model.", require_prefix=False):
     """
     Read state dict as fake tensors
     """
     reader = gguf.GGUFReader(path)
     sd = {}
     dt = {}
+    arch_field = reader.get_field("general.architecture")
+    if arch_field is not None:
+        if len(arch_field.types) != 1 or arch_field.types[0] != gguf.GGUFValueType.STRING:
+            raise TypeError(f"Bad type for GGUF general.architecture key: expected string, got {arch_field.types!r}")
+        arch_str = str(arch_field.parts[arch_field.data[-1]], encoding="utf-8")
+        if arch_str not in {"flux", "sd1", "sdxl"}:
+            raise ValueError(f"Unexpected architecture type in GGUF file, expected one of flux, sd1, sdxl but got {arch_str!r}")
+    else:
+        arch_str = None
+    if handle_prefix is not None:
+        prefix_len = len(handle_prefix)
+        tensor_names = set(tensor.name for tensor in reader.tensors)
+        has_prefix = any(s.startswith(handle_prefix) for s in tensor_names)
+    else:
+        has_prefix = False
+    if require_prefix and not has_prefix:
+        raise RuntimeError(f"Loader requires prefix {handle_prefix!r} which does not exist in the model")
     for tensor in reader.tensors:
-        tensor_name = str(tensor.name)
+        sd_key = tensor_name = tensor.name
+        if has_prefix:
+            if not tensor_name.startswith(handle_prefix):
+                continue
+            sd_key = tensor_name[prefix_len:]
+        tensor_type_str = str(tensor.tensor_type)
         torch_tensor = torch.from_numpy(tensor.data) # mmap
         shape = gguf_sd_loader_get_orig_shape(reader, tensor_name)
         if shape is None:
             shape = torch.Size(tuple(int(v) for v in reversed(tensor.shape)))
-        elif tensor.tensor_type in {gguf.GGMLQuantizationType.F32, gguf.GGMLQuantizationType.F16}:
+            if arch_str is None and (tensor_name.endswith(".proj_in.weight") or tensor_name.endswith(".proj_out.weight")):
+                # Workaround for stable-diffusion.cpp SDXL detection.
+                # Impossible to land here for our own GGUF files as we set architecture.
+                while len(shape) > 2 and shape[-1] == 1:
+                    shape = shape[:-1]

```

Comment:
this isn't amazingly elegant but it will only have an effect when loading `stable-diffusion.cpp` GGUF files. they don't store an architecture so there's no way to know exactly what type of model we're dealing with without adding heuristics.

if we are ever loading a non-SDXL `stable-diffusion.cpp` format GGUF file which is incompatible squeezing those dimensions that then this will have to be reevaluated.

Reply:
Looks good enough to me as a workaround for now. As you say, we can always deal with it later if it comes to it.